### ISSUE TITLE
failure of "float nastiness: half bounded generates endpoint"

### ISSUE THREAD

**Issue by bukzor** [HypothesisWorks/hypothesis#1266]
I've been running into this problem: https://paste.pound-python.org/show/qchV5aFbMkvs0uoYfArd/

This particular test has a @flaky decorator but no @given decorator. Despite the decorator, the problem was happening 100% reproducibly for me, and upon debugging, I found that the stdlib random singleton was in a deterministic state. Further, when running the test alone (`-k "half and endpoint"`), the test passed, and no longer had determinism.

This is caused by cross-test pollution. The @given decorator results in `random.seed(0)` running, so that the prior test causes the next test to be deterministic even though it wouldn't normally.

I'm not sure of the best fix, but this patch fixes the problem for me, and obviates the @flaky decorator too:

```
--- b/hypothesis-python/tests/common/debug.py
+++ a/tests/common/debug.py
@@ -73,6 +73,10 @@
         definition, condition=None,
         settings=None, random=None
 ):
+    if random is None:
+      from random import seed
+      seed(0)
+
     settings = Settings(
         settings,
         max_examples=10000,
```

**Comment by Zac-HD** [comment]
I think the proper solution will avoid causing this reseeding pollution in the first place.  For example, a new context manager that can be used in `core.py`:

```python
@contextlib.context_manager
def deterministic_context(seed=0):
    state = random.getstate()
    random.seed(seed)
    yield
    rangom.setstate(state)
```

And then ban direct use of `random.seed` to avoid similar problems in future.

**Comment by bukzor** [comment]
> I think the proper solution will avoid causing this reseeding pollution in the first place.

To be clear, there's no **re**seeding involved in the problem above. The test does no seeding at all, but other tests in the module do, so it has differing behavior when run alone or along with all tests in the module.

**Comment by Zac-HD** [comment]
I mean that the other tests should not have the side-effect of changing the `random` state, which can be implemented via the decorator I suggested.

**Comment by Zac-HD** [comment]
Note: fixing this issue will remove the last reason to use the `random_module` strategy, which should therefore be turned into a no-op and deprecated.

**Comment by bukzor** [comment]
@Zac-HD Is there a ticket for that idea? How would I generate a pareto-random number in that world? Right now I can use st.randoms() and random.pareto().

**Comment by Zac-HD** [comment]
This issue is the ticket!  I've assigned myself to make that clearer 😄 

In that world you can use `st.randoms()` and methods on the instance it gives you (as now), *or* just use the global random - Hypothesis will pick a seed for each example and take care of resetting the global random state after each example.

**Comment by DRMacIver** [comment]
> Note: fixing this issue will remove the last reason to use the random_module strategy, which should therefore be turned into a no-op and deprecated.

Wait, what? It might remove *our* last use of it, but the `random_module` strategy serves a perfectly valid (if somewhat niche) use case. Why would a change about how we test our code remove that?

**Comment by Zac-HD** [comment]
Sorry, looks like I need to actually write out my chain of logic here 😊 

- The failure is *caused* by the way that every test with `@given` includes a `random.seed(0)` in `core.py`, but never resets the PRNG to its previous state.  This massively degrades the quality of randomness in any test suite using Hypothesis, which is *very bad*.  VERY VERY BAD.  See e.g. #1174 for an example where this has bitten us 😭.  We should fix the general problem rather than patching the tests to avoid it, *especially* because this impacts users too.

- The `random_module()` strategy is a special magic thing that doesn't give you a value (users: "*what???*"), but instead draws an integer seed and does the whole save/seed/resetting thing for you.  Sadly, this only helps if you (a) know it exists, (b) know you need it for a particular test, and (c) remember to use it.

- So we need to do a thing where we save and reset the PRNG state in `@given` (and IIRC `find`, and for state machines - see #702) to avoid polluting the randomness of the rest of the test suite.  If we do this, the damage is contained to each test where `random_module()` should have been used, but wasn't.  🎉 

Before doing some deeper investigation I had hoped it would be practical to integrate the `random_module()` functionality into internals, but adding seeding to the mix makes it a lot harder.

**Comment by Zac-HD** [comment]
I'll open a pull to reset state in a few days when life calms down a little :smile: 

We agree on the actual purpose of `random_module` being to vary the PRNG state between examples.  However, it *does* reset the state in cleanup... right back the the `random.seed(0)` state that we set up earlier, which causes the original problem! https://github.com/HypothesisWorks/hypothesis/blob/d522674b77a90ca02d87293dedebc731497c5b4e/hypothesis-python/src/hypothesis/strategies.py#L1002-L1009

**Comment by DRMacIver** [comment]
> However, it does reset the state in cleanup... 

Yes, sorry, that's what I meant by "the `random.seed(0)` is called beforehand". i.e. `random_module` does exactly as much state cleanup as without it effectively (less even! At least the state of the system after can vary a bit with normal `@given`)
